pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -101,13 +108,52 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(proxy *model.Proxy, push *mo
 		clusters = append(clusters, patcher.insertedClusters()...)
 	}
 
-	return cb.normalizeClusters(clusters)
+	for _, c := range clusters {
+		resources = append(resources, &discovery.Resource{Name: c.Name, Resource: util.MessageToAny(c)})
```

Comment:
resources here contain cached as well as non cached clusters correct? Are n't we caching Any already? If yes, we do not have to call util.MessageToAny() for cached clusters?

Reply:
`resources` contains already marshalled to Any. This is returned by outbound cluster builder because we have to cache the any.

The other builders which are not yet doing caching are just returning the original Cluster, so now we marshal them.

Reply:
So the outbound clusters are marshalled twice - I do not know the impact of it, but is it possible to avoid it like adding them to cachedresources and add cachedresources as is at the end?

Reply:
They are not marshalled twice.

For outbound (with caching): we return them as `Any` and add to `resources`.
For other types (without caching): we return them as `Cluster`, add them to `clusters`, then marshal and add to `resources`.

Reply:
Makes Sense. Sorry I missed in the code that outbound clusters are being directly added to resources.